lib/logflare/backends.ex

```diff
@@ -478,21 +479,62 @@ defmodule Logflare.Backends do
     end
   end
 
+  def push_recent_events(%Source{}, []), do: :ok
+
+  def push_recent_events(%Source{id: source_id}, log_events) do
+    existing = :ets.lookup(:recent_events, source_id)
+
+    :ets.delete(:recent_events, source_id)
+
+    # sort newest to oldest
+    objects =
+      for event <- Enum.sort_by(log_events, fn %{body: %{"timestamp" => ts}} -> ts end, &<=/2) do
+        # don't cache source information
+        {source_id, %{event | source: nil}}
+      end
+
+    :ets.insert(:recent_events, Enum.take(objects ++ existing, 100))
+
+    :ok
+  end
+
   @doc """
   Lists the latest recent logs of all caches across the cluster.
-  Performs a check to ensure that the cache is started. If not started yet globally, it will start the cache locally.
   """
-  @spec list_recent_logs(Source.t()) :: [LogEvent.t()]
-  def list_recent_logs(%Source{} = source) do
-    RecentLogsServer.list_for_cluster(source.token)
+  @spec list_recent_events(Source.t()) :: [LogEvent.t()]
+  def list_recent_events(%Source{} = source) do
+    {results, _} = Cluster.Utils.rpc_multicall(__MODULE__, :list_recent_events_local, [source])
+
+    List.flatten(results)
+    |> Enum.sort_by(fn %{body: %{"timestamp" => ts}} -> ts end, &<=/2)
+    |> Enum.take(-100)
   end
 
   @doc """
   Lists latest recent logs of only the local cache.
   """
-  @spec list_recent_logs_local(Source.t()) :: [LogEvent.t()]
-  def list_recent_logs_local(%Source{} = source) do
-    RecentLogsServer.list(source.token)
+  @spec list_recent_events_local(Source.t()) :: [LogEvent.t()]
+  def list_recent_events_local(%Source{id: source_id}) do
+    ms =
+      Ex2ms.fun do
+        {id, event} when id == ^source_id -> event
+      end
+
+    :ets.select(:recent_events, ms)
+    |> Enum.sort_by(fn %{body: %{"timestamp" => ts}} -> ts end, &<=/2)
```

Comment:
So we need to sort here if we're sorting in insert?

Reply:
I guess unless it's a `sorted_set` we should make sure to re-sort.

Reply:
Yea we can remove this since it's an ordered_set